### DISCUSSION TITLE
Custom add_metric function in time_series.compare_models()

### DISCUSSION THREAD

**Discussion by tbellamey** [pycaret/pycaret#2842] [Q&A] [answered]
Hi PyCaret community,

I'm trying to create a custom add_metric function for usage in time_series.compare_modules()

Functionally, I would like this new add_metric function - named 'OPE' for 'Overall Percentage Error' - to calculate the absolute aggregated percentage error over a multi-step forecast horizon. 

Here's an example of the math logic I'd like to implement, using 3 periods of a multi-step forecast:

- **y**
  - y_1 = 10
  - y_2 = 7
  - y_3 = 16

- **y_pred**
  - y_pred_1 = 8
  - y_pred_2 = 11
  - y_pred_3 = 11
 
- **sum(y)**
  - y_1 + y_2 + y_3 = 10 + 7 + 16 = 33
 
- **sum(y_pred)**
  - y_pred_1 + y_pred_2 + y_pred_3 = 8 + 11 + 11 = 30
 
- **OPE**
  - abs(1 - sum(y)/sum(y_pred) ) = abs(1 - 33/30) = abs(1 - 1.1) = abs(-0.1) = **0.10** 

However, when I run compare_models() with the code below, I receive nan values for the OPE scores. 

**Defining the OPE score function**
`def ope_score_func(y,y_pred,**kwargs):
    return abs(1-(float(np.sum(y))/float(np.sum(y_pred))))`

**Adding the new OPE score function to the TSForecastingExperiment() by invoking add_metric**
`ts_exp.add_metric(id='ope',name='OPE',score_func=ope_score_func,greater_is_better=False)`

**Running compare_models() to generate a leaderboard sorted by OPE score function**
`best = ts_exp.compare_models(turbo=True, sort='OPE')`

The compare_models() function completes the call & does provide a leaderboard of models with standard metric scores. 

But in the new OPE column, all of the model scores return as nan values. 

I assume I am not utilizing the parameters correctly within the custom ope_score_func() code. 
But based on PyCaret's readthedocs, I did confirm that this custom function should have the correct signature: 
**score_func(y, y_pred, \*\*kwargs)**

Any tips to resolve?

**Comment by ngupta23** [comment, marked as answer]
See if this helps. If it does not, please file an Issue with a reproducible example

https://gist.github.com/ngupta23/410152d2104c73d814cd25b934be11bb

**Reply by tbellamey** [reply]
Thanks Nikhil, this example helped - resolved!

**Reply by ngupta23** [reply]
Glad it helped. It would help others if you can post the issue and the solution here. Thanks!

**Reply by annmariajoseph** [reply]
I am facing the same issue .I m trying to add metric “profit” to compare_models
The function i used is for prediction profit for customer churn but the column is showing only 0 value after adding 
This is my function

def calcluate_profit(y_true,y_pred):
Tp=no.where(y_pred==1) &(y_true==1),(5000-1000),0)
Fp=np.where((y_pred==1)&(y_true==0),-1000,0)
return no.sum(Tp)+np.sum(fp)

**Comment by AM764733** [comment]
I am trying to add wmape and smape to my regression problem which is essentially sort of multivariate timeseries framed as regression - this is how i am doing it for now                                                                                                                   
 def wmape(y_true, y_pred):
    """
    Calculates the Weighted Mean Absolute Percentage Error.
    """
    epsilon = 1e-8 
    abs_percentage_error = (np.abs(y_true - y_pred) / (np.abs(y_true) + epsilon)) * 100
    weighted_ape = abs_percentage_error * y_true
    return np.sum(weighted_ape) / np.sum(y_true)

def smape(y_true, y_pred):
    """
    Calculates the Symmetric Mean Absolute Percentage Error.
    """
    epsilon = 1e-8 
    return 100/len(y_true) * np.sum(2 * np.abs(y_pred - y_true) / (np.abs(y_true) + np.abs(y_pred) + epsilon))        
                                  
exp = setup() #some setup   
exp.add_metric("wmape", "WMAPE", wmape, greater_is_better=False)
    exp.add_metric("smape", "SMAPE", smape, greater_is_better=False)

The problem i am having is that i am getting 0 value for all WMAPE and SMAPE, Can anyone help me with this?

**Reply by celestinoxp** [reply]
with the recent changes to scikit-learn 1.4, it is possible that the pycaret metrics file needs to be revised... you can help fix errors in the code... take a look at [https://github.com/pycaret/pycaret/blob/master/pycaret/internal/metrics.py](https://github.com/pycaret/pycaret/blob/master/pycaret/internal/metrics.py)

help fix the pycaret code by making a pull-request
